### PR TITLE
Ruleset stats v3

### PR DESCRIPTION
**Pull request by glongo**
A correct rebase of #1720.

Ticket: [#1585] (https://redmine.openinfosecfoundation.org/issues/1585)

Prscript:
- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/86
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/85

### CODE REVIEW

File: src/detect.c [hunk 1]
```diff
@@ -292,6 +292,69 @@ char *DetectLoadCompleteSigPath(const DetectEngineCtx *de_ctx, char *sig_file)
...
+
+    sig->filename = SCStrdup(sig_file);
+    if (sig->filename == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error allocating memory");
+        return 0;
+    }
+
+    sig->sig_str = SCStrdup(sig_str);
+    if (sig->sig_str == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error allocating memory");
```
Comment (inliniac) [inline src/detect.c]: sig->filename is lost

File: src/output-json-stats.c [hunk 2]
```diff
@@ -65,6 +67,90 @@ typedef struct JsonStatsLogThread_ {
...
+static TmEcode OutputTenancyStats2Json(int output_type, json_t **jdata)
+{
+    *jdata = json_object();
+    if (*jdata == NULL) {
+        return TM_ECODE_FAILED;
+    }
+
+    DetectEngineCtx *list = DetectEngineGetCurrent();
+    if (list == NULL) {
+        return TM_ECODE_FAILED;
```
Comment (inliniac) [inline src/output-json-stats.c]: jdata is lost

File: src/output-json-stats.c [hunk 3]
```diff
@@ -65,6 +67,90 @@ typedef struct JsonStatsLogThread_ {
...
+        json_object_set_new(jdata, "rules_loaded",
+                            json_integer(sig_stat->good_sigs_total));
+        json_object_set_new(jdata, "rules_failed",
+                            json_integer(sig_stat->bad_sigs_total));
+    }
+
+    return jdata;
+}
+
+static TmEcode OutputTenancyStats2Json(int output_type, json_t **jdata)
```
Comment (inliniac) [inline src/output-json-stats.c]: func has multiple mem leak conditions

Comment (inliniac) [inline src/output-json-stats.c]: why not have the function return a json object or NULL in case of error?

File: src/output-json-stats.c [hunk 4]
```diff
@@ -120,6 +216,19 @@ static int JsonStatsLogger(ThreadVars *tv, void *thread_data, const StatsTable *
...
         json_integer((int)difftime(tval.tv_sec, st->start_time)));
 
+    /* tenancy stats */
+    if (DetectEngineMultiTenantEnabled()) {
+        json_t *js_tenancy = json_object();
+        if (js_tenancy == NULL) {
+            return 0;
+        }
+
+        TmEcode retval = OutputTenancyStats2Json(BOTH, &js_tenancy);
```
Comment (inliniac) [inline src/output-json-stats.c]: unclear what BOTH means here. Please rename it to something meaningful

File: src/output-json-stats.h [hunk 5]
```diff
@@ -24,6 +24,16 @@
 #ifndef __OUTPUT_JSON_COUNTERS_H__
 #define __OUTPUT_JSON_COUNTERS_H__
 
+enum {
```
Comment (inliniac) [inline src/output-json-stats.h]: meaningless without comment

File: src/output-json-stats.c [hunk 6]
```diff
@@ -83,6 +169,16 @@ static json_t *OutputStats2Json(json_t *js, const char *key)
     json_t *value = json_object_iter_value(iter);
     if (value == NULL) {
         value = json_object();
+        /* When multi-tenancy is disabled, we put engine stats
```
Comment (inliniac) [inline src/output-json-stats.c]: can you expand on this logic? I still don't get it

File: src/unix-manager.c [hunk 7]
```diff
@@ -647,18 +649,104 @@ TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
...
+    SCEnter();
+    TmEcode retval = TM_ECODE_OK;
+    json_t *jdata = NULL;
+
+    retval = OutputTenancyStatsLastReload(&jdata);
+    if (jdata == NULL) {
+        json_object_set_new(server_msg, "message", json_string("Unable to get info"));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    if (retval == TM_ECODE_FAILED) {
```
Comment (inliniac) [inline src/unix-manager.c]: this condition leads to exactly the same result as when it's omitted

File: src/unix-manager.c [hunk 8]
```diff
@@ -647,18 +649,104 @@ TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
...
+    SCEnter();
+    json_t *jdata = NULL;
+    TmEcode retval = TM_ECODE_OK;
+
+    retval = OutputTenancyStatsRuleset(&jdata);
+    if (jdata == NULL) {
+        json_object_set_new(server_msg, "message", json_string("Unable to get info"));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    if (retval == TM_ECODE_FAILED) {
```
Comment (inliniac) [inline src/unix-manager.c]: unnecessary code block, w/o it the effect is the same

File: src/unix-manager.c [hunk 9]
```diff
@@ -647,18 +649,104 @@ TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
...
+
+    json_object_set_new(server_msg, "message", jdata);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+TmEcode UnixManagerShowFailedRules(json_t *cmd,
+                                   json_t *server_msg, void *data)
+{
+    SCEnter();
+    DetectEngineCtx *de_ctx = DetectEngineGetCurrent();
```
Comment (inliniac) [inline src/unix-manager.c]: this takes a reference of de_ctx, which means after you're done with it you need to DetectEngineDeReference

File: src/unix-manager.c [hunk 10]
```diff
@@ -647,18 +649,104 @@ TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
...
+    SigString *sigs_str = de_ctx->sig_stat.failed_sigs;
+    json_t *js_sigs_array = json_array();
+
+    if (js_sigs_array == NULL) {
+        json_object_set_new(server_msg, "message", json_string("Unable to get info"));
```
Comment (inliniac) [inline src/unix-manager.c]: js_sigs_array is lost

File: src/output-json-stats.c [hunk 11]
```diff
@@ -65,6 +67,90 @@ typedef struct JsonStatsLogThread_ {
...
+}
+
+static TmEcode OutputTenancyStats2Json(int output_type, json_t **jdata)
+{
+    *jdata = json_object();
+    if (*jdata == NULL) {
+        return TM_ECODE_FAILED;
+    }
+
+    DetectEngineCtx *list = DetectEngineGetCurrent();
```
Comment (inliniac) [inline src/output-json-stats.c]: takes reference, need to deref later

File: src/detect.c [hunk 12]
```diff
@@ -292,6 +292,69 @@ char *DetectLoadCompleteSigPath(const DetectEngineCtx *de_ctx, char *sig_file)
     return path;
 }
 
+static SigString *SigStringAlloc(void)
+{
+    SigString *sigstr = SCMalloc(sizeof(SigString));
```
Comment (inliniac) [inline src/detect.c]: I prefer SCCalloc as then the structure is initialized to 0. Now only 2 of the 4 fields are initialized.

**Comment by inliniac** [comment]
Needs more work.

File: src/detect.c (hunk 12 above)
Comment (glongo) [inline src/detect.c]: ack

File: src/detect.c (hunk 1 above)
Comment (glongo) [inline src/detect.c]: ack

File: src/output-json-stats.c (hunk 3 above)
Comment (glongo) [inline src/output-json-stats.c]: ack

File: src/output-json-stats.c (hunk 11 above)
Comment (glongo) [inline src/output-json-stats.c]: ack

File: src/output-json-stats.c (hunk 2 above)
Comment (glongo) [inline src/output-json-stats.c]: ack

File: src/output-json-stats.c (hunk 6 above)
Comment (glongo) [inline src/output-json-stats.c]: ack

File: src/output-json-stats.c (hunk 4 above)
Comment (glongo) [inline src/output-json-stats.c]: ack

File: src/output-json-stats.h (hunk 5 above)
Comment (glongo) [inline src/output-json-stats.h]: ack

File: src/unix-manager.c (hunk 7 above)
Comment (glongo) [inline src/unix-manager.c]: ack

File: src/unix-manager.c (hunk 8 above)
Comment (glongo) [inline src/unix-manager.c]: ack

File: src/unix-manager.c (hunk 9 above)
Comment (glongo) [inline src/unix-manager.c]: ack

File: src/unix-manager.c (hunk 10 above)
Comment (glongo) [inline src/unix-manager.c]: ack

**Pull request closed without merging**